src/mgmt/srv_target.c

```diff
@@ -784,92 +715,146 @@ ds_mgmt_hdlr_tgt_create(crt_rpc_t *tc_req)
 	D_ASSERT(tc_in != NULL && tc_out != NULL);
 
 	/** insert record in dpt_creates_ht hash table (creates in progress) */
-	D_ALLOC_PTR(ptrec);
-	if (ptrec == NULL) {
-		D_ERROR("failed to alloc ptrec\n");
+	D_ALLOC_PTR(tca.tca_ptrec);
+	if (tca.tca_ptrec == NULL)
 		D_GOTO(out_reply, rc = -DER_NOMEM);
-	}
-	uuid_copy(ptrec->dptr_uuid, tc_in->tc_pool_uuid);
-	ptrec->cancel_create = false;
+	uuid_copy(tca.tca_ptrec->dptr_uuid, tc_in->tc_pool_uuid);
+	tca.tca_ptrec->cancel_create = false;
 	ABT_mutex_lock(pooltgts->dpt_mutex);
-	rc = d_hash_rec_insert(&pooltgts->dpt_creates_ht, ptrec->dptr_uuid,
-			       sizeof(uuid_t), &ptrec->dptr_hlink, true);
+	rc = d_hash_rec_insert(&pooltgts->dpt_creates_ht,
+			       tca.tca_ptrec->dptr_uuid, sizeof(uuid_t),
+			       &tca.tca_ptrec->dptr_hlink, true);
 	ABT_mutex_unlock(pooltgts->dpt_mutex);
 	if (rc == -DER_EXIST) {
 		D_ERROR(DF_UUID": already creating or cleaning up\n",
 			DP_UUID(tc_in->tc_pool_uuid));
-		rc = -DER_AGAIN;
-		goto out_rec;
+		D_GOTO(out_rec, rc = -DER_AGAIN);
 	} else if (rc) {
 		D_ERROR(DF_UUID": failed insert dpt_creates_ht: "DF_RC"\n",
 			DP_UUID(tc_in->tc_pool_uuid), DP_RC(rc));
 		goto out_rec;
 	}
 	D_DEBUG(DB_MGMT, DF_UUID": record inserted to dpt_creates_ht\n",
-		DP_UUID(ptrec->dptr_uuid));
-
-	/** generate path to the target directory */
-	rc = ds_mgmt_tgt_file(tc_in->tc_pool_uuid, NULL, NULL, &path);
-	if (rc)
-		D_GOTO(out, rc);
+		DP_UUID(tca.tca_ptrec->dptr_uuid));
 
-	/** check whether the target already exists */
-	rc = access(path, F_OK);
-	if (rc >= 0) {
-		/** target already exists, let's reuse it for idempotence */
-		/** TODO: fetch tgt uuid from existing DSM pool */
-		uuid_generate(tgt_uuid);
-
-		/**
-		 * flush again in case the previous one in tgt_create()
-		 * failed
-		 */
-		rc = dir_fsync(path);
-	} else if (errno == ENOENT) {
-		/** target doesn't exist, create one */
-		rc = tgt_create(ptrec, tc_in->tc_pool_uuid, tgt_uuid,
-				tc_in->tc_scm_size, tc_in->tc_nvme_size, path);
-	} else {
+	tca.tca_scm_size  = tc_in->tc_scm_size;
+	tca.tca_nvme_size = tc_in->tc_nvme_size;
+	rc = pthread_create(&thread, NULL, tgt_create_preallocate, &tca);
```

Comment:
Just curious, could you use ABT_thread_create() to make sure this thread on non-0 xstream?  Then you could use ABT condition to notify the success of the thread, instead of those pthread join/cancel.

Reply:
ABT_thread is ULT. So, in case of blocking I/O whole HW thread will be blocked with all ULTs which scheduled on them.

Reply:
I mean schedule this ULT on a non-SWIM and less critical xstream, for example xstream 1 .